### PR TITLE
Add support for EP to context parallelism in self-attention

### PR DESCRIPTION
**Pull request by shuningjin**
# Description

**Goal**

For mixture of expert models, we may use expert parallelism. For attention layer, EP acts as FSDP currently. Built upon [previous context parallelism work](https://github.com/AI-Hypercomputer/maxtext/pull/1445), this PR is to introduce the option of using EP as CP for attention. This is joint effort with @RissyRan.

FIXES: b/418396648

**Main code changes**
- `attentions.py`
  - MHA, MLA, tpu_flash_attention
  - changed logical_axis

- `base.yml`
  - change logical_axis_rules
  - add a flag to customize expert sharding behavior in attention `expert_shard_attention_option`:  fsdp or context

- unit test: `tests.attention_test`
  - `AttentionTest.test_tpu_flash_attention_cp_and_ep` &  `MLATest.test_tpu_flash_attention_cp_and_ep` (extended from [cp test](https://screenshot.googleplex.com/499pjxu8HiV8z23))
  - using cp/ep with tpu_flash_attention, with or without context_parallel_load_balance
  - compare logit against dot_product without sharding
 

**Use case**
- training, MoE with MHA / MLA attention
- cp_load_balance={true, false}, tpu_flash_attention
- Example (ep_as_cp): `ici_expert_parallelism=4, ici_context_parallelism=1, expert_shard_attention_option=context`, shard context by 4 in attention, shard expert by 4 for moe
- Example (ep_as_cp + native cp): `ici_expert_parallelism=2, ici_context_parallelism=2, expert_shard_attention_option=context`, shard context by 4 in attention, shard expert by 2 and context by 2 for moe

# Tests

Tested on v5p-8

Verify sharding shape
- end-to-end pretraining: reduced version of Mixtral-8x7b (for MHA) and reduced DeepSeek3-671b (for MLA)
- tpu_flash_kernel, `context_parallel_load_balance=True`, parallelism: `ici_expert_parallelism=2, ici_context_parallelism=2, expert_shard_attention_option=context` and `ici_expert_parallelism=2, ici_context_parallelism=2, expert_shard_attention_option=fsdp`
- See test details in  b/418396648#comment7

Verify attention output logit against dot product
- use the added unit test 
- tpu_flash_kernel, attention {MHA, MLA}, `context_parallel_load_balance={True, False}`, parallelism: {`ici_expert_parallelism=4, expert_shard_attention_option=context`, `ici_expert_parallelism=2, expert_shard_attention_option=context, ici_context_parallelism=2`} 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -403,6 +407,11 @@ ici_autoregressive_parallelism: 1
 ici_pipeline_parallelism: 1
 ici_expert_parallelism: 1
 
+# How to use ici_expert_pallelism for attention sharding
+# "fsdp" (ep acts as fsdp parallelism)
+# "context" (ep acts as context parallelism, training only)
+expert_shard_attention_option: "fsdp"
```
Comment (RissyRan) [inline MaxText/configs/base.yml]: Shall we put this near MoE configs? This will be only needed for MoE models. I doubt if dcn_expert_parallelism is working out of box (we just haven't tested yet). So update `How to use expert pallelism` instead?

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -710,6 +719,9 @@ def tpu_flash_attention(
     """TPU Flash Attention."""
 
     cp_size = self.mesh.shape["context"]
+    if self.config.expert_shard_attention_option == "context":
```
Comment (RissyRan) [inline MaxText/layers/attentions.py]: Nit: let's put `context` as variable to avoid typo in the future?

File: MaxText/layers/moe.py [hunk 3]
```diff
@@ -949,7 +939,7 @@ def generate_masks_subgroup(self, top_k_indices, softmax_probs):
     )
     expert_token_count = nn.with_logical_constraint(
         expert_token_count,
-        ("activation_batch", "activation_length", None, None, None),
+        ("activation_batch", "activation_norm_length", None, None, None),
```
Comment (RissyRan) [inline MaxText/layers/moe.py:963]: Thanks for keeping them aligned!

File: MaxText/pyconfig.py [hunk 4]
```diff
@@ -162,6 +170,7 @@ def validate_keys(keys):
   validate_model_call_mode(keys["model_call_mode"])
   validate_prefill_and_target_lengths(keys["max_prefill_predict_length"], keys["max_target_length"])
   validate_rope_type(keys["rope_type"])
+  validate_expert_shard_attention_option(keys["expert_shard_attention_option"])
```
Comment (RissyRan) [inline MaxText/pyconfig.py]: Let's validate them inside of `if keys["num_experts"] > 1` for MoE models

File: MaxText/tests/attention_test.py [hunk 5]
```diff
@@ -560,48 +551,103 @@ def tpu_kernel_attention_helper(self, num_kv_heads):
         jax.numpy.allclose(mha_generic_output, mha_generic_flash_output, rtol=1e-01, atol=1e-01, equal_nan=False)
     )
 
+  @pytest.mark.tpu_only
+  def test_tpu_flash_attention_cp_and_ep(self):
+    # cp
+    self.tpu_flash_attention_cp_and_ep_helper(
+        ici_context_parallelism=4,  # use context parallelism of 4
```
Comment (RissyRan) [inline MaxText/tests/attention_test.py]: Nit: probably we could clean up those comments? I think config's name present its purpose? Similar to other comments bellow.

File: MaxText/tests/attention_test.py [hunk 6]
```diff
@@ -560,48 +551,103 @@ def tpu_kernel_attention_helper(self, num_kv_heads):
         jax.numpy.allclose(mha_generic_output, mha_generic_flash_output, rtol=1e-01, atol=1e-01, equal_nan=False)
     )
 
+  @pytest.mark.tpu_only
+  def test_tpu_flash_attention_cp_and_ep(self):
+    # cp
```
Comment (RissyRan) [inline MaxText/tests/attention_test.py]: You could comment something like `ep -> fsdp, load_balance=false` for each test.

**Review by RissyRan — COMMENTED** [review]
Thanks Shuning! Great work!

File: MaxText/configs/base.yml (hunk 1 above)
Comment (shuningjin) [reply MaxText/configs/base.yml]: Thanks for reviewing! I have updated my code to address all your comments.

File: MaxText/layers/attentions.py (hunk 2 above)
Comment (shuningjin) [reply MaxText/layers/attentions.py]: Done

File: MaxText/layers/moe.py (hunk 3 above)
Comment (shuningjin) [reply MaxText/layers/moe.py:953]: Done

File: MaxText/pyconfig.py (hunk 4 above)
Comment (shuningjin) [reply MaxText/pyconfig.py]: Done

File: MaxText/tests/attention_test.py (hunk 5 above)
Comment (shuningjin) [reply MaxText/tests/attention_test.py]: Done

File: MaxText/tests/attention_test.py (hunk 6 above)
Comment (shuningjin) [reply MaxText/tests/attention_test.py]: Done

File: MaxText/configs/base.yml [hunk 7]
```diff
@@ -160,6 +160,11 @@ tile_batch_seq: 512
 tile_activation_dim: 1024
 tile_weight_dim: 1024
 
+# How to use expert parallelism for attention sharding
```
Comment (gobbleturk) [inline MaxText/configs/base.yml]: I would change to "How the expert axis is used to shard attention weights and activations"

File: MaxText/pyconfig.py [hunk 8]
```diff
@@ -200,6 +208,7 @@ def validate_keys(keys):
     validate_sparse_matmul_parallelism(keys)
     validate_ragged_dot(keys)
     validate_deepseek_moe(keys)
+    validate_expert_shard_attention_option(keys["expert_shard_attention_option"])
```
Comment (RissyRan) [inline MaxText/pyconfig.py:211]: This only needed when there are experts right? could we move inside of `if keys["num_experts"] > 1:`?

**Review by RissyRan — APPROVED** [review]

**Review by gobbleturk — APPROVED** [review]

File: MaxText/configs/base.yml [hunk 9]
```diff
@@ -308,18 +313,22 @@ logical_axis_rules: [
                       ['activation_embed_and_logits_batch', ['data', 'stage', 'fsdp', 'fsdp_transpose', 'expert']],
                       ['activation_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence','autoregressive']],
                       ['activation_kv_heads', ['tensor', 'tensor_transpose', 'sequence','tensor_sequence']],
-                      ['activation_length', ['sequence', 'context']],
-                      ['prefill_activation_length', ['sequence', 'context']],
-                      ['activation_length', ['context']],
+                      ['activation_length', ['sequence', 'context', 'expert']],
+                      ['activation_length', ['context', 'expert']],
+                      ['activation_length_no_exp', ['sequence', 'context']],
```
Comment (richjames0) [inline MaxText/configs/base.yml:327]: Matt and I think this would be nicer named as e.g. activation_length_mlp and activation_length_attn

File: MaxText/layers/attentions.py [hunk 10]
```diff
@@ -710,6 +719,9 @@ def tpu_flash_attention(
     """TPU Flash Attention."""
 
     cp_size = self.mesh.shape["context"]
+    if self.config.expert_shard_attention_option == EP_AS_CONTEXT:
```
Comment (richjames0) [inline MaxText/layers/attentions.py]: Could this be done in pyconfig?

File: MaxText/layers/moe.py (hunk 3 above)
Comment (richjames0) [reply MaxText/layers/moe.py:963]: I half get why these were different but I think it's better to make them the same. That said I think `activation_length` is the more general and so I would use that name. Norm relates specifically to an RMSNorm which may or may not be present

File: MaxText/tests/attention_test.py [hunk 11]
```diff
@@ -560,48 +551,105 @@ def tpu_kernel_attention_helper(self, num_kv_heads):
         jax.numpy.allclose(mha_generic_output, mha_generic_flash_output, rtol=1e-01, atol=1e-01, equal_nan=False)
     )
 
+  @pytest.mark.tpu_only
```
Comment (richjames0) [inline MaxText/tests/attention_test.py:577]: I'd have to look it up but it's possible to parameterize pytest fixtures with different values - that would be quite a bit cleaner. Also Matt has concerns around how long these will take - did you check that?

**Review by richjames0 — COMMENTED** [review]
Really impressive that you understood this and got it working!

File: MaxText/configs/base.yml (hunk 7 above)
Comment (shuningjin) [reply MaxText/configs/base.yml]: Done

File: MaxText/pyconfig.py (hunk 8 above)
Comment (shuningjin) [reply MaxText/pyconfig.py:211]: This is already inside `if keys["num_experts"] > 1:`

File: MaxText/tests/attention_test.py (hunk 11 above)
Comment (shuningjin) [reply MaxText/tests/attention_test.py:577]: TLDR: I have updated AttentionTest from `unittest.TestCase` to `parameterized.TestCase`, to be consistent with MLATest. The CP-related tests are parameterized now.

- I tried `@pytest.mark.parametrize`. It is not compatible since AttentionTest is `unittest.TestCase`
- Another possibility is to use `from parameterized import parameterized, @parameterized.expand` for `unittest.TestCase`. Yet for consistency, I choose the following option.
- I noticed for MLATest, Gagik uses `from absl.testing import parameterized, parameterized.TestCase, parameterized.named_parameters`.Therefore, I updated AttentionTest from  `unittest.TestCase` to `parameterized.TestCase` (which inherits from`unittest.TestCase`), and used `parameterized.named_parameters`.

File: MaxText/configs/base.yml (hunk 9 above)
Comment (shuningjin) [reply MaxText/configs/base.yml:327]: We are actually using both `activation_length` and `activation_length_no_exp` in attention. https://screenshot.googleplex.com/9ddYFMoS3pwNWqr

File: MaxText/layers/moe.py (hunk 3 above)
Comment (shuningjin) [reply MaxText/layers/moe.py:963]: `activation_norm_length` was previously used in many other layers. We are aligning towards them.
https://screenshot.googleplex.com/62RSWJR46SbozFY

File: MaxText/tests/attention_test.py (hunk 11 above)
Comment (shuningjin) [reply MaxText/tests/attention_test.py:577]: Total test time is 94.49s on v5p-8. `AttentionTest` and `MLATest` each contains 6 parameterized cases. 
```
24.82s call     MaxText/tests/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel_cp_ep_no_load_balance
11.14s call     MaxText/tests/attention_test.py::MLATest::test_tpu_flash_attention_context_parallel_cp_ep_no_load_balance
8.60s call     MaxText/tests/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel_ep_no_load_balance
8.29s call     MaxText/tests/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel_cp_no_load_balance
6.32s call     MaxText/tests/attention_test.py::MLATest::test_tpu_flash_attention_context_parallel_ep_no_load_balance
6.31s call     MaxText/tests/attention_test.py::MLATest::test_tpu_flash_attention_context_parallel_cp_no_load_balance
4.24s call     MaxText/tests/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel_ep_with_load_balance
4.22s call     MaxText/tests/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel_cp_with_load_balance
4.19s call     MaxText/tests/attention_test.py::AttentionTest::test_tpu_flash_attention_context_parallel_cp_ep_with_load_balance
3.99s call     MaxText/tests/attention_test.py::MLATest::test_tpu_flash_attention_context_parallel_ep_with_load_balance
3.62s call     MaxText/tests/attention_test.py::MLATest::test_tpu_flash_attention_context_parallel_cp_with_load_balance
3.61s call     MaxText/tests/attention_test.py::MLATest::test_tpu_flash_attention_context_parallel_cp_ep_with_load_balance

=== 12 passed, 22 deselected in 94.49s (0:01:34)
```
Should this be further reduced? @gobbleturk

File: MaxText/layers/attentions.py (hunk 10 above)
Comment (shuningjin) [reply MaxText/layers/attentions.py]: Done. 
- I have updated the pyconfig.py. Adding `raw_keys["context_parallel_size"] = get_context_parallel_size(raw_keys)`: context_parallel_size is either `ici_context * dcn_context`  or `ici_context * dcn_context * ici_expert * dcn_expert` depends on config.expert_shard_attention_option
- also updated the places using context_parallel_size: here in flash attention, train.py, and  attention_test.py

**Comment by shuningjin** [comment]
Resolved merge conflict with nnx migration for attention layer
- https://github.com/AI-Hypercomputer/maxtext/pull/2066
- https://github.com/AI-Hypercomputer/maxtext/pull/2089

Re-testing on local v5p-8, diff (before vs. after nnx migration)
- Mini mixtral: [1.1](https://diff.googleplex.com/#key=AgwFAaH1mYdr), [1.2](https://diff.googleplex.com/#key=wDVkYJhFbYW8). Mini deepseek3: [2.1](https://diff.googleplex.com/#key=2YujHauLFfo9), [2.2](https://diff.googleplex.com/#key=YDkhaGxX59zC)
- Sharding shape is still correct, training loss is close
- TFLOP/s/device is better now. In both cases, the FLOPs calculation has included [the recent change](https://github.com/AI-Hypercomputer/maxtext/pull/1988). Possible reasons: different docker image (jax 0.6.2 vs. 0.7.0), linen vs. nnx.
